Doc/library/typing.rst

```diff
@@ -1230,6 +1232,83 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    Usage::
+
+        Ts = TypeVarTuple('Ts')
+        Shape = TypeVarTuple('Shape')
+
+    Vanilla type variables enable parameterization with a *fixed* number of
+    types. Type variable tuples, on the other hand, allow parameterization with
+    an *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        class Array(Generic[*Shape]):
+            pass
+
+        from typing import NewType
+        Height = NewType('Height', int)
+        Width = NewType('Width', int)
```

Comment:
IMO the use of `NewType` is just distracting. You can just state that both `Array[int]` and `Array[int, int]` are valid, and define 1d and 2d arrays, respectively. Maybe you could add a matching definition of `__getitem__` -- would this work?
```py
def __getitem__(self, key: Shape) -> float: ...
```
(to keep it simple, hardcoding the return type).

Reply:
Yeah, fair. The `__getitem__` example is nice, but I'd prefer to keep it out of the very first example block, because I think the information it imparts (how to use TypeVarTuples with argument annotations; the fact that TypeVarTuples used in methods refer to the same parameters as used to instantiate the class) are not the _primary_ thing we want to impart (that `Array` can take an arbitrary number of type parameters), so I've used it in the second example block.